### PR TITLE
QK logits clipping (non-split version)

### PR DESCRIPTION
**Pull request by BoxiangW**
Added MLA and MHA(GQA) clipping support

### CODE REVIEW

**Comment by BoxiangW** [comment]
TE's https://github.com/NVIDIA/TransformerEngine/pull/2195 (2.9.0) is needed for this PR

**Comment by skyw** [comment]
> TE's [NVIDIA/TransformerEngine#2195](https://github.com/NVIDIA/TransformerEngine/pull/2195) (2.9.0) is needed for this PR

It has been merged.

**Comment by skyw** [comment]
Boxiang, I suppose you need someone in the expert reviewer list to review.

File: megatron/core/extensions/transformer_engine.py [hunk 1]
```diff
@@ -989,6 +989,12 @@ def __init__(
             self.kept_packed_seq_params.discard("cu_seqlens_q_padded")
             self.kept_packed_seq_params.discard("cu_seqlens_kv_padded")
 
+        # qk-clip is only supported in TE 2.9.0 and later
+        if config.qk_clip and is_te_min_version("2.9.0"):
+            # TE 2.9.0 introduces return_max_logit for qk-clip getting the max attention logits
+            extra_kwargs["return_max_logit"] = True
+            self.current_max_attn_logits = None
```
Comment (yaoyu-33) [inline megatron/core/extensions/transformer_engine.py:1256]: what if `not is_te_min_version` but config.qk_clip? might raise an error

File: megatron/core/optimizer/qk_clip.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    Args:
+        model: The model to clip the QK attention logits, a list of model chunks.
+
+    Returns:
+        The maximum attention logit, a float.
+    """
+
+    log_max_attention_logit = 0
+    for model_chunk in model:
```
Comment (yaoyu-33) [inline megatron/core/optimizer/qk_clip.py]: i dont remember, is model always a list?

Comment (BoxiangW) [reply megatron/core/optimizer/qk_clip.py]: According to https://github.com/NVIDIA/Megatron-LM/pull/1929/files#diff-7fc2a91343e49be3c1705b8237bbe437029132d8826e5c963e1b0b84d457bc38L1250. I think it is true for MLM, is it true for MBridge as well?

File: megatron/core/extensions/transformer_engine.py (hunk 1 above)
Comment (BoxiangW) [reply megatron/core/extensions/transformer_engine.py:1256]: Right now it will raise an error if te version is wrong

**Comment by deepakn94** [comment]
Can we re-name this PR? It should just be "QK logits clipping" or something similar?

File: megatron/core/optimizer/qk_clip.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+
+import torch
+
+from megatron.core import mpu
```
Comment (jaredcasper) [inline megatron/core/optimizer/qk_clip.py]: Should use parrallel_state, mpu is a legacy name.

File: megatron/core/optimizer/qk_clip.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    with torch.no_grad():
+        log_max_attention_logit = 0
+        for model_chunk in model:
+            for transformer_layer in model_chunk.module.module.decoder.layers:
+                if hasattr(transformer_layer.self_attention, 'clip_qk'):
+                    torch.distributed.all_reduce(
+                        transformer_layer.self_attention.core_attention.current_max_attn_logits,
+                        op=torch.distributed.ReduceOp.MAX,
+                        group=mpu.get_data_parallel_group(with_context_parallel=True),
```
Comment (jaredcasper) [inline megatron/core/optimizer/qk_clip.py]: Is there any way to get this group in via args? We are trying to move away from global process groups.

File: megatron/core/optimizer/qk_clip.py (hunk 3 above)
Comment (BoxiangW) [reply megatron/core/optimizer/qk_clip.py]: Thanks for catching this, changed to `parrallel_state` instead
